### ISSUE TITLE
Authenticated authorization bypass in `tsf_crop_image` allows users with `upload_files` to crop arbitrary attachments

### ISSUE THREAD

**Issue by HuajiHD** [HuajiHD/CVE#11]
## Product

The SEO Framework

## Affected Version

Verified on 5.1.4

Likely affects versions containing the current `crop_image()` implementation introduced in 5.0.0.

## Severity

Medium

## CWE

- CWE-862: Missing Authorization
- CWE-639: Authorization Bypass Through User-Controlled Key

## Summary

The AJAX handler `wp_ajax_tsf_crop_image` authorizes requests with a generic `upload_files` capability check and a generic nonce, but it does not enforce object-level authorization on the supplied attachment ID.

As a result, a logged-in user who has `upload_files` and access to a TSF media-enabled edit screen can submit an arbitrary attachment ID that they are not allowed to edit, and the plugin will crop and copy that image and create a new attachment entry derived from the unauthorized source attachment.

This issue was dynamically reproduced in a local Docker-based WordPress environment with an Author account targeting an Administrator-owned attachment.

## Technical Details

The vulnerable flow is:

- `bootstrap/init-admin-ajax.php`
  - registers `add_action( 'wp_ajax_tsf_crop_image', [ AJAX::class, 'crop_image' ] );`
- `inc/classes/admin/script/ajax.class.php`
  - `crop_image()` calls `Utils::check_ajax_capability_referer( 'upload_files' )`
  - reads `$_POST['id']` into `$attachment_id`
  - passes the ID to `wp_crop_image()`
  - copies metadata from the original attachment
  - creates a new attachment with `wp_insert_attachment()`
  - returns the new attachment via `wp_send_json_success()`
- `inc/classes/admin/script/utils.class.php`
  - `create_ajax_capability_nonce()` creates a nonce bound only to the capability string
  - `check_ajax_capability_referer()` validates only `current_user_can( 'upload_files' )` and that generic nonce
- `inc/classes/admin/script/loader.class.php`
  - exposes the `upload_files` nonce to TSF media-enabled edit screens via `tsfMediaL10n.nonce`

The handler never verifies whether the current user may edit or manage the specific source attachment, for example with `current_user_can( 'edit_post', $attachment_id )`.

## Code References

- [[bootstrap/init-admin-ajax.php#L38-L40](https://github.com/mnt/d/cve/the-seo-framework-5.1.4/bootstrap/init-admin-ajax.php#L38)](/mnt/d/cve/the-seo-framework-5.1.4/bootstrap/init-admin-ajax.php#L38)
- [[inc/classes/admin/script/ajax.class.php#L149-L166](https://github.com/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/ajax.class.php#L149)](/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/ajax.class.php#L149)
- [[inc/classes/admin/script/ajax.class.php#L189-L230](https://github.com/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/ajax.class.php#L189)](/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/ajax.class.php#L189)
- [[inc/classes/admin/script/ajax.class.php#L252](https://github.com/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/ajax.class.php#L252)](/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/ajax.class.php#L252)
- [[inc/classes/admin/script/utils.class.php#L83-L112](https://github.com/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/utils.class.php#L83)](/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/utils.class.php#L83)
- [[inc/classes/admin/script/loader.class.php#L665](https://github.com/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/loader.class.php#L665)](/mnt/d/cve/the-seo-framework-5.1.4/inc/classes/admin/script/loader.class.php#L665)
- [[lib/js/media.js#L404-L412](https://github.com/mnt/d/cve/the-seo-framework-5.1.4/lib/js/media.js#L404)](/mnt/d/cve/the-seo-framework-5.1.4/lib/js/media.js#L404)

## Prerequisites

- The victim attachment is an image attachment
- The attacker is authenticated
- The attacker has `upload_files`
- The attacker can access a TSF media-enabled post edit screen and obtain the TSF `upload_files` nonce

## Reproduction Steps

1. Log in as an Administrator and create or upload an image attachment. In my test, the original attachment ID was `9`.

2. Log in as an Author user.

3. Verify the Author cannot edit the Administrator-owned attachment directly:

   `GET /wp-admin/post.php?post=9&action=edit`

   Result: `403` and "Sorry, you are not allowed to edit this item."

4. As the Author, open one of the Author's own post edit screens where TSF media scripts are loaded.

5. Extract `tsfMediaL10n.nonce` from the page source or browser devtools.

6. Send the following POST request to `/wp-admin/admin-ajax.php` using the Author session:

```http
POST /wp-admin/admin-ajax.php HTTP/1.1
Host: <target>
Content-Type: application/x-www-form-urlencoded

action=tsf_crop_image&
nonce=<tsf_upload_files_nonce>&
id=9&
context=tsf-image&
cropDetails[x1]=0&
cropDetails[y1]=0&
cropDetails[width]=64&
cropDetails[height]=64&
cropDetails[dst_width]=64&
cropDetails[dst_height]=64
```

7. Observe that the request succeeds and returns a new attachment ID.

## Observed Result

The Author could not edit the original attachment directly, but the AJAX request still succeeded and created a new cropped attachment derived from that original image.

Example result from my reproduction:

- Original attachment ID: `9`
- New attachment ID: `10`

Dynamic reproduction output:

```json
{
  "author_cannot_edit_original_attachment": true,
  "ajax_success": true,
  "original_attachment_id": 9,
  "new_attachment_id": 10
}
```

## Expected Result

The request should be rejected unless the current user is authorized to manage the specific source attachment referenced by the supplied attachment ID.

## Impact

This is an object-level authorization bypass on media attachments.

A lower-privileged authenticated user can:

- target arbitrary image attachment IDs
- server-side crop and copy images they are not authorized to manage
- create a new attachment record derived from the unauthorized source image
- inherit associated metadata such as title, description, caption, and alt text from the original attachment

This can expose restricted editorial assets and breaks attachment ownership and authorization boundaries.

## Suggested Remediation

Before calling `wp_crop_image()`, enforce object-level authorization on the supplied attachment ID. For example:

- validate that the target is an attachment image
- require `current_user_can( 'edit_post', $attachment_id )`

A minimal fix would be:

```php
if ( ! current_user_can( 'edit_post', $attachment_id ) ) {
	wp_send_json_error(
		[ 'message' => __( 'Unauthorized request.', 'autodescription' ) ],
		403
	);
}
```

Optionally, the nonce model can also be tightened so it is not bound only to the generic `upload_files` capability.

## Environment Used for Dynamic Verification

- Date verified: 2026-03-23
- WordPress: 6.x in local Docker lab
- Plugin: The SEO Framework 5.1.4
- Administrator account owning the source attachment
- Author account used as the attacker
- Local lab path: [[compose.yml](https://github.com/mnt/d/cve/.tmp-the-seo-framework-514-repro/compose.yml)](/mnt/d/cve/.tmp-the-seo-framework-514-repro/compose.yml)
- Local PoC: [[repro-crop-bypass.py](https://github.com/mnt/d/cve/.tmp-the-seo-framework-514-repro/docker/repro-crop-bypass.py)](/mnt/d/cve/.tmp-the-seo-framework-514-repro/docker/repro-crop-bypass.py)

## Notes

The dynamic lab required normalizing write permissions on the Docker uploads directory because the `wordpress` and `wordpress:cli` images use different `www-data` UIDs. That was an environment-specific Docker issue and does not affect the validity of the authorization bypass itself.
